src/diff_generate.c

```diff
@@ -645,8 +645,11 @@ int git_diff__oid_for_entry(
 		if (fd < 0)
 			error = fd;
 		else {
+			uint64_t file_size;
+			if ((error = git_futils_filesize(&file_size, fd)) < 0)
+				file_size = (uint64_t)entry.file_size;
 			error = git_odb__hashfd_filtered(
-				out, fd, (size_t)entry.file_size, GIT_OBJECT_BLOB, fl);
+				out, fd, (size_t)file_size, GIT_OBJECT_BLOB, fl);
```

Comment:
I think that I would take the opportunity to invert the `oid_for_entry` / `oid_for_file` interaction.

Right now, if we call `oid_for_file` with the actual known size of the file (as a `git_object_size_t`), then we discard the high bits to stuff that into a `git_index_entry` and call `oid_for_entry`.  You're proposing doing another `stat` to look up the filesize.

Instead, let's make `oid_for_entry` call `oid_for_file` - any existing callers of `oid_for_file` can work unchanged (and accurately, since they don't cast away the high bits of the size).  If we really need to look up the file size on disk when we have a `git_index_entry` (since it is only 32 bits by definition) then we can do it only for them.

However, looking a bit more closely, I don't actually know why `git_odb__hashfd_filtered` _takes_ a size.  It doesn't need it.  In fact, I'm a little skeptical that there aren't some bugs here when we _do_ give it a size.  I haven't done an audit but you can imagine a scenario where you pass the size to this function, and it reads exactly _up to that size_ and then stops and returns the hash of the data _up to that point_ even though there's still more bytes in that file descriptor.

This is not a big deal if we always stat and read - it's a race condition, sure, but that's rather undefined.  My bigger concern would be if we take a `git_index_entry` out of the index, and assume that we should read `file_size` bytes.  Maybe we don't do that and my concerns are unfounded, but if we want the OID for a file on-disk, I'm thinking that there's no reason not to just _read the whole file on disk_ and not take a size parameter at all.

Reply:
I like the idea of inverting the `oid_for_entry` / `oid_for_file` interaction. I implemented that idea in the updated pull request. This causes more calls to `stat` when computing the `oid_for_entry`. One could optimise this by first checking if the `git_index_entry` might be exceeding 32 bits, what do you think?

I am not sure if I understand 100% of your reasoning and what other implications my changes might have regarding `git_odb__hashfd_filtered`. Frankly speaking I am not feeling confident refactoring that part of the code.